### PR TITLE
gh-113190: Reenable non-debug interned string cleanup

### PR DESCRIPTION
**Pull request by eduardo-elizondo**
In GH-19474 we introduced immortalized interned strings. However, at the time, we didn't have a strict guarantee that a runtime shutdown (i.e `Py_Finalize()`) would completely cleanup all the allocated PyObjects (i.e a simple example of `./python -X showrefcount -c 'import itertools'` would show leaks).

Today, these leaks have been largely fixed and there have been stricter guarantees into how obmalloc and the interpreter state works after a runtime shutdown. Given this, it should be safe to re-enable the interned string cleanup and revert the structseq test that had issues with this leaks before. This test should show the correctness of this code now.

* Issue: gh-113190
* Issue: gh-113055

### CODE REVIEW

**Comment by Eclips4** [comment]
I added `DO-NOT-MERGE` for you :)

**Comment by eduardo-elizondo** [comment]
> I added `DO-NOT-MERGE` label for you :)

Thanks!! Will ping you once all the tests are passing to move this into safe-to-merge mode :)

**Comment by eduardo-elizondo** [comment]
@Eclips4 tests seem to be passing! Could you help me by adding the "skip news" label (since it's probably not required) and removing the "DO-NOT-MERGE" label?

Thanks!!

**Comment by eduardo-elizondo** [comment]
cc @nascheme to take a look when you have a chance!

**Comment by nascheme** [comment]
It would be nice to do a better cleanup but I'm concerned about how safe this actually is.  Since we don't have an accurate refcnt for these strings, we are assuming that no one is actually still using them.  It seems possible that some extensions or programs embedding Python could hang on to interned strings after `Py_Finalize()`.  That would risk introducing "use-after-free" bugs.  In the worst case, that could turn into an arbitrary code execution security bug.

Maybe I'm being too paranoid?   I don't know how we could actually fix this though.  Keeping hold of and using memory allocated by the Python runtime after `Py_Finalize()`  surely puts you an very shaky ground.  Previously, the obmalloc state was global to the process and you could keep using memory allocated from there after finalizing.   With sub-interpeters potentially having their own obmalloc state, that's likely not okay.

My gut feeling is we should proceed cautiously, even though we would like to fix the leak.  Maybe we could enable this for alphas and then turn it off again for release?   Maybe we need to wait until we get `python -X showrefcount` showing zero for all or nearly all built-in and well known extensions?  Seems like a tall order though.

**Comment by eduardo-elizondo** [comment]
@nascheme I agree with what you mentioned above and I was paranoid of that scenario too, hence the disabling of the code.

Purely going off from the docs, I don't think there's anywhere where Python provides an explicit expectation of what the behavior is under the presence of a memory leak after a call to Py_Finalize. Furthermore, in the C-API it is specified that [`Py_Finalize`](https://docs.python.org/3.13/c-api/init.html#c.Py_FinalizeEx) will: "Undo all initializations made by [Py_Initialize()](https://docs.python.org/3.13/c-api/init.html#c.Py_Initialize) and subsequent use of Python/C API functions". So this is technically doing that for all calls to `PyUnicode_New`.

This gives an an implicit assumption that holding onto these interned strings after Py_Finalize is an undefined behavior. Perhaps we should be more explicit about this in the docs (in the "Bugs and Caveats" section of Py_Finalize)? i.e python can't guarantee correctness for any memory allocated during a python initialization and used in any subsequent initializations. Thoughts?

Regardless, I completely agree that we should still be careful about putting this out there. Stating with alphas sounds like a good idea to get some early signal. Let me know what's the best way to only enable this only for alphas.

**Comment by vstinner** [comment]
> Maybe we need to wait until we get python -X showrefcount showing zero for all or nearly all built-in and well known extensions?

It's mostly done for one year: https://mail.python.org/archives/list/python-dev@python.org/thread/E4C6TDNVDPDNNP73HTGHN5W42LGAE22F/ "Mostly" means: some remaining stdlib C extensions still implement their own static types which are not cleared by `_PyTypes_FiniTypes()`, and/or don't use the multi-phase initialization API yet.

`test_embed` checks that `python -c pass` does not leak memory at Python exit: see `MiscTests.test_no_memleak()`.

> Purely going off from the docs, I don't think there's anywhere where Python provides an explicit expectation of what the behavior is of memory that survives a call to Py_Finalize.

This change is a backward incompatible. I just to document well the change in What's New in Python 3.13:

* Explain how to qualify if a crash is related to this issue,
* Document how to update affected code.

In short, a program must not keep references to any Python object between `Py_Finalize()` and `Py_Initialize()`. IMO the general rule is calling `Py_Finalize()` converts all Python object pointers to dangling pointers. There are some very specific exceptions which should be listed with conditions, such as "only on CPython and it's an implementation detail which can change between two Python versions".

In general, I suggest to reset all pointers to Python objects at each Py_Initialize() call. Or said differently: don't store *any* pointer to Python objects after Py_Finalize() :-)

Examples of exceptions:

* Immortal objects
* Static types (built-in immortal objects are marked as immortal)
* Singletons (built-in singleton objects are marked as immortal) such as: empty string, empty tuple, small integers (-5..256).

**Comment by eduardo-elizondo** [comment]
>  This change is a backward incompatible. I just to document well the change in What's New in Python 3.13

For sure, I can update the PR with the details with what you mentioned here.

> It's mostly done for one year

Given this guarantee, I think it's safe to assume that if we have any Python object leak it's a user induced leak, right? i.e. the runtime itself will not cause incorrect behavior. I will include some wording around this so that users can isolate the surface are when looking for issues (i.e look at the third-party extensions not at the runtime).

> In short, a program must not keep references to any Python object between Py_Finalize() and Py_Initialize()

That's my take too - though it's not specified in the docs! Any thoughts of also updating the Py_Finalize/Py_Initialize documentation as part of this PR to include stricter language like this?

**Comment by vstinner** [comment]
> This change is a backward incompatible. I suggest to just document well the change in What's New in Python 3.13

I looked again at the code. It has a complicated history:

* In Python 3.9 and older, interned strings were **never deleted**. Only special build for Valgrind and Purity cleared them at exit.
* In Python 3.10 and 3.11, interned strings are **always deleted** at exit.
* In Python 3.12, interned strings are **deleted only if Python is built in debug mode**: they are not deleted in release mode.

If there are applications relying on interned strings to survive between multiple Py_Initialize()/Py_Finalize() cycles, they should already be affected by Python 3.10 and 3.11.

By the way, Python 3.11 added _PyTypes_FiniTypes() to "clear" static types in Py_Finalize().

**Comment by eduardo-elizondo** [comment]
@nascheme @vstinner I've updated the PR to add better language in the documentation based on our conversations - happy to revise the wording or phrasing that I've included. In any case, let me know if this looks good to go

File: Misc/NEWS.d/next/Core and Builtins/2024-01-15-18-11-48.gh-issue-113190.OwQX64.rst [hunk 1]
```diff
@@ -0,0 +1,20 @@
+This updates the interned string deallocation function
+``_PyUnicode_ClearInterned`` to delete all interned strings during runtime
+finalization when calling ``Py_Finalize``, regardless of their reference
+count.
+
+Worth noting that if an extension accidentally holds onto a interned string,
+event after calling Py_Finalize, it will result in use-after-free error,
```
Comment (nascheme) [inline Misc/NEWS.d/next/Core and Builtins/2024-01-15-18-11-48.gh-issue-113190.OwQX64.rst]: Should be "even after calling ...", I think.

**Comment by nascheme** [comment]
The extra comments and docs are a good improvement.  I think this can be merged.   If we find it breaks too many extension modules, it might be that we have to revert it.  However, given Victor's comments on the history, that seems somewhat unlikely since those extensions probably didn't work with older versions of Python.

File: Misc/NEWS.d/next/Core and Builtins/2024-01-15-18-11-48.gh-issue-113190.OwQX64.rst (hunk 1 above)
Comment (eduardo-elizondo) [reply Misc/NEWS.d/next/Core and Builtins/2024-01-15-18-11-48.gh-issue-113190.OwQX64.rst]: Good catch, fixed!

**Comment by eduardo-elizondo** [comment]
Sounds good, fixed the typos so the PR is ready to go! Once it's merged, I'll be monitoring things closely in case we get complaints or bug reports.

**Comment by neonene** [comment]
I think this PR is ready to be merged, as PEP-687 is about to be completed. The only concern would be the third-party extension modules with single-phase init.

cc @eduardo-elizondo @ericsnowcurrently @nascheme

**Comment by neonene** [comment]
Could anyone disable the `backport to 3.12` option?  Thanks.  cc @Eclips4

**Comment by ericsnowcurrently** [comment]
FYI, given https://github.com/python/cpython/issues/113993#issuecomment-2147420017, this PR may become irrelevant.

**Review by encukou — DISMISSED** [review]
#120520 will unfortunately conflict with this.
Please do not merge this one first though. I can help resolve the conflicts later.

File: Doc/whatsnew/3.14.rst [hunk 2]
```diff
@@ -368,6 +368,17 @@ New Features
 
   (Contributed by Victor Stinner in :gh:`119182`.)
 
+* Modified the ``_PyUnicode_ClearInterned`` function to always delete all
+  interned strings during a call to :c:func:`Py_Finalize`. This
```
Comment (encukou) [inline Doc/whatsnew/3.14.rst]: These are public docs, they shouldn't mention internal functions.
```suggestion
* :c:func:`Py_Finalize` now deletes all interned strings. This
```

File: Misc/NEWS.d/next/Core and Builtins/2024-01-15-18-11-48.gh-issue-113190.OwQX64.rst [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+For example, in Python 3.9 and older, interned strings were never deleted.
+Only special builds for Valgrind and Purity cleared them at exit. In Python
+3.10 and 3.11, interned strings are always deleted at exit. In Python 3.12,
+interned strings are deleted only if Python is built in debug mode: they
+are not deleted in release mode. In Python 3.13, interned strings will
+now be all deleted.
+
+Given how the guarantees changed throughout different versions, it is not
+expected that users actively rely on this behavior for their extensions.
```
Comment (encukou) [inline Misc/NEWS.d/next/Core and Builtins/2024-01-15-18-11-48.gh-issue-113190.OwQX64.rst]: The blurb should be short.

```suggestion
:c:func:`Py_Finalize` now deletes all interned strings.
```

The warning would be useful in porting notes (What's New); the reasons why the change can be made is best kept in the issue, PR, or commit message.

**Comment by encukou** [comment]
@eduardo-elizondo Do you want to finish the docs here, or should I take over?

**Comment by eduardo-elizondo** [comment]
@encukou updated and addressed the comments, this should be ready to go!

File: Doc/whatsnew/3.14.rst [hunk 4]
```diff
@@ -417,6 +417,16 @@ New Features
   which has an ambiguous return value.
   (Contributed by Irit Katriel and Erlend Aasland in :gh:`105201`.)
 
+* :c:func:`Py_Finalize` now deletes all interned strings. This
+  is backwards incompatible to any C-Extension that holds onto an interned
+  string after a call to :c:func:`Py_Finalize` and is then reused after a
+  call to :c:func:`Py_Initialize`.  Any issues arising from this behavior will
+  normally result in crashes during the exectuion of the subsequent call to
+  :c:func:`Py_Initialize` from accessing uninitialized memory. To fix, use
+  an address sanitizer (i.e. ASAN) to identify any use-after-free coming from
```
Comment (encukou) [inline Doc/whatsnew/3.14.rst]: I don't think the acronym is useful here. It's not used again, and I don't think searching for “ASAN” will get you much more than “address sanitizer”.
```suggestion
  an address sanitizer to identify any use-after-free coming from
```
If you want to keep it, [we prefer](https://devguide.python.org/documentation/style-guide/#use-simple-language) English terms (“that is”) over Latin abbreviation like “i.e.”.

**Review by encukou — APPROVED** [review]
Looks great, thank you! I'll run the refleak buildbots as there might have been regressions.

A style nit:

**Comment by eduardo-elizondo** [comment]
Thanks @encukou !!

**Comment by neonene** [comment]
Is there any chance this will be backported to 3.13.1?

**Comment by hugovk** [comment]
The conflict preventing Miss Islington from backporting https://github.com/python/cpython/pull/113601 was the addition to `Doc/whatsnew/3.14.rst`.

I've made the backport PRs. Do we need any other docs in them?

**Comment by encukou** [comment]
> Is there any chance this will be backported to 3.13.1?

Sorry, I missed the notification!

See this paragraph in the added docs:

>   Note that Python will do a best effort at freeing all memory allocated by the Python
>   interpreter.  Therefore, any C-Extension should make sure to correctly clean up all
>   of the preveiously allocated PyObjects before using them in subsequent calls to
>   :c:func:`Py_Initialize`.  Otherwise it could introduce vulnerabilities and incorrect
>   behavior.

In other words, it's possible that some extensions will start crashing after `Py_Finalize` & `Py_Initialize`. 
It would only be extensions that used the API incorreclty, but, that's little consolation for affected users.

I don't think we should backport to bugfix branches, unfortunately.

**Comment by hugovk** [comment]
> I don't think we should backport to bugfix branches, unfortunately.

Okay, I'll close the backports.

**Pull request merged**